dagstore/dagstorewrapper.go

```diff
@@ -0,0 +1,219 @@
+	io.Closer
+}
+
+type dagStoreWrapper struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+
+	dagStore DagStore
+	mountApi LotusMountAPI
+}
+
+var _ DagStoreWrapper = (*dagStoreWrapper)(nil)
+
+func NewDagStoreWrapper(cfg MarketDAGStoreConfig, mountApi LotusMountAPI) (*dagStoreWrapper, error) {
+	// construct the DAG Store.
+	registry := mount.NewRegistry()
+	if err := registry.Register(lotusScheme, NewLotusMountTemplate(mountApi)); err != nil {
+		return nil, xerrors.Errorf("failed to create registry: %w", err)
+	}
+
+	failureCh := make(chan dagstore.ShardResult, 1)
+	dcfg := dagstore.Config{
+		TransientsDir: cfg.TransientsDir,
+		IndexDir:      cfg.IndexDir,
+		Datastore:     cfg.Datastore,
+		MountRegistry: registry,
+		FailureCh:     failureCh,
+	}
+	dagStore, err := dagstore.NewDAGStore(dcfg)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to create DAG store: %w", err)
+	}
+
+	return NewDagStoreWrapperWithDeps(dagStore, mountApi, failureCh)
+}
+
+// Used by the tests
+func NewDagStoreWrapperWithDeps(dagStore DagStore, mountApi LotusMountAPI, failureCh chan dagstore.ShardResult) (*dagStoreWrapper, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	dw := &dagStoreWrapper{
+		ctx:    ctx,
+		cancel: cancel,
+
+		dagStore: dagStore,
+		mountApi: mountApi,
+	}
+
+	dw.wg.Add(1)
+	// the dagstore will write Shard failures to the `failureCh` here. Run a go-routine to handle them.
+	go dw.handleFailures(failureCh)
+
+	return dw, nil
+}
+
+func (ds *dagStoreWrapper) handleFailures(failureCh chan dagstore.ShardResult) {
+	defer ds.wg.Done()
+	ticker := time.NewTicker(gcInterval)
+	defer ticker.Stop()
+
+	select {
+	case <-ticker.C:
+		_, _ = ds.dagStore.GC(ds.ctx)
+	case f := <-failureCh:
+		log.Errorw("shard failed", "shard-key", f.Key.String(), "error", f.Error)
+		if err := ds.dagStore.RecoverShard(ds.ctx, f.Key, nil, dagstore.RecoverOpts{}); err != nil {
+			log.Warnw("shard recovery failed", "shard-key", f.Key.String(), "error", err)
+		}
+	case <-ds.ctx.Done():
+		return
+	}
+}
+
+func (ds *dagStoreWrapper) LoadShard(ctx context.Context, pieceCid cid.Cid) (carstore.ClosableBlockstore, error) {
+	key := shard.KeyFromCID(pieceCid)
+	resch := make(chan dagstore.ShardResult, 1)
+	err := ds.dagStore.AcquireShard(ctx, key, resch, dagstore.AcquireOpts{})
+
+	if err != nil {
+		if xerrors.Unwrap(err) != dagstore.ErrShardUnknown {
+			return nil, xerrors.Errorf("failed to schedule acquire shard for piece CID %s: %w", pieceCid, err)
+		}
+
+		// if the DAGStore does not know about the Shard -> register it and then try to acquire it again.
+		log.Infow("ErrShardUnknown during LoadShard, will re-register", "pieceCID", pieceCid)
+		if err := RegisterShardSync(ctx, ds, pieceCid, "", false); err != nil {
+			return nil, xerrors.Errorf("failed to re-register shard during loading piece CID %s: %w", pieceCid, err)
+		}
+		log.Infow("Successfully re-registered Shard in LoadShard", "pieceCID", pieceCid)
```

Comment:
nit: Also, remove code references from logging lines. Log lines are intended for operators, and operators don't usually read the code. Prefer:

```
log.Infow("successfully re-registered shard", "pieceCID", pieceCid)
```

In fact, the logging output already prints out the log site as director/file:line.

Reply:
@raulk This can happen if the upfront shard migration on the Storage Provider after the upgrade fails for whatever reason. I remember we'd discussed some rare but hairy edge cases there that can cause this.

I agree with your recommendations here and am incorporating them in the the upcoming PR.

Reply:
Fixed in https://github.com/filecoin-project/go-fil-markets/pull/577.